x-pack/packages/ml/query_utils/src/filter_based_default_query.ts

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { isPopulatedObject } from '@kbn/ml-is-populated-object';
+
+import type { FilterBasedSimpleQuery } from './types';
+import { isDefaultQuery } from './default_query';
+
+const boolRequiredAttributes = ['filter', 'must', 'must_not'];
+
+// should identify variants of
+// `{ bool: { filter: [{ match_all: {} }], must: [], must_not: [], should: [] } }`
+// `{ bool: { filter: [], must: [{ match_all: {} }], must_not: [] } }`
+export function isFilterBasedDefaultQuery(arg: unknown): arg is FilterBasedSimpleQuery {
+  return (
+    isPopulatedObject(arg, ['bool']) &&
+    isPopulatedObject(arg.bool, boolRequiredAttributes) &&
+    Object.keys(arg.bool).every(
+      // should be either an empty array or an array with just 1 default query
+      (d) => {
+        if (!isPopulatedObject(arg.bool, [d])) return false;
```

Comment:
Unless I'm misunderstanding this, I don't think this check is needed.
We know `d` must exist in `arg.bool` because we've used `Object.keys`?

If this is the case and this check isn't needed, then this loop could use `Object.entries`.

Reply:
You're right, this is not necessary and more complex than needed; I changed it to just use `Object.values` because we don't need the `key` at all. Updated in [aa908b2](https://github.com/elastic/kibana/pull/176534/commits/aa908b262d8de1cbf7e788a57d99ff69f1f35bbc).